core/domain/opportunity_services.py

```diff
@@ -386,3 +386,218 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
+    new SkillOpportunityModel otherwise.
+
+    Args:
+        topic_id: str. The topic_id of the to-be-added topic.
+        skill_id: str. The skill_id of the SkillOpportunityModel.
+    """
+    skill_opportunity_models = (
+        opportunity_models.SkillOpportunityModel.get_by_skill_id(skill_id))
+    topic = topic_fetchers.get_topic_by_id(topic_id)
+    for model in skill_opportunity_models:
+        if model.topic_id == topic_id:
+            return
+        if model.topic_id is None:
+            model.topic_id = topic_id
+            model.topic_name = topic.name
+            model.put()
+            return
+
+    if skill_opportunity_models:
+        create_skill_opportunity(
+            opportunity_domain.SkillOpportunity(
+                topic_id=topic_id,
+                topic_name=topic.name,
+                skill_id=skill_id,
+                skill_description=skill_opportunity_models[0].skill_description
+            )
+        )
+
+
+def update_skill_opportunities_with_topic_name(topic_id, topic_name):
+    """Updates the SkillOpportunityModel(s) with a new topic_name.
+
+    Args:
+        topic_id: str. The corresponding topic id of the opportunity.
+        topic_name: str. The new topic name.
+    """
+    skill_opportunity_models = (
+        opportunity_models.SkillOpportunityModel.get_by_topic_id(topic_id))
+    for model in skill_opportunity_models:
+        model.topic_name = topic_name
+        model.put()
+
+
+def delete_topic_from_skill_opportunity(topic_id, skill_id):
+    """Deletes topic_id and topic_name from SkillOpportunityModel(s).
+
+    Args:
+        topic_id: str. The topic_id of the to-be-deleted topic.
+        skill_id: str. The skill_id of the SkillOpportunityModel.
+    """
+    skill_opportunity_models = (
+        opportunity_models.SkillOpportunityModel.get_by_skill_id(skill_id))
+    for model in skill_opportunity_models:
+        if model.topic_id == topic_id:
+            model.topic_id = None
+            model.topic_name = None
+            model.put()
+
+
+def delete_topic_from_all_skill_opportunities(topic_id):
+    """Deletes topic_id and topic_name from SkillOpportunityModel(s).
+
+    Args:
+        topic_id: str. The topic_id of the to-be-added topic.
+    """
+    skill_opportunity_models = (
+        opportunity_models.SkillOpportunityModel.get_by_topic_id(topic_id))
+    for model in skill_opportunity_models:
+        model.topic_id = None
+        model.topic_name = None
+        model.put()
+
+
+def update_skill_opportunity_question_counts(skill_ids, delta):
+    """Increments question_count(s) of SkillOpportunityModel(s) with
+    corresponding skill_ids.
+
+    Args:
+        skill_ids: list(str). A list of skill_ids corresponding to
+            SkillOpportunityModel(s).
+        delta: int. The delta for which to update each question_count.
+    """
+    for skill_id in skill_ids:
+        _update_skill_opportunity_question_count(skill_id, delta)
```

Comment:
Try to refactor so that you do a single put_multi() in the operation.

Similar comments apply elsewhere.

Reply:
Done.